### PR TITLE
Add JAX train/predict support

### PR DESCRIPTION
**Pull request by trivoldus28**
Notes:
- BufferPredict is factored out of the Tensorflow `Predict` and then
used as a parent class for JAX `Predict`. This should make using
BufferPredict easier for other implementations.
- `Train` was written based off PyTorch implementation. Since JAX is
rather low level and allowing many freedom for model implementation,
to make things easier for us, we make a simple `GenericJaxModel`
interface for models to follow in order to train or predict. A model
implementing this interface will need to contain not only the forward
model but also loss and update fn. Some examples can be found in
https://github.com/funkelab/funlib.learn.jax

Pre-Merge Checklist:

- [ ] if this PR adds a feature: request merge into the latest development branch (`vX.Y-dev`)
- [ ] if this PR fixes a bug: request merge into the latest patch branch (`patch-X.Y.Z`)
- [ ] PR branch name is short and describes the feature/bug addressed in this PR
- [ ] commit messages [are consistent](https://chris.beams.io/posts/git-commit/)
- [ ] changes reviewed by another contributor
- [ ] tests cover changes
- [ ] all tests pass

### CODE REVIEW

File: gunpowder/jax/nodes/train.py [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+        for output_name, array_key in self.outputs.items():
+            if array_key in request:
+                array_outputs[array_key] = output_name
+
+        return array_outputs
+
+    def __collect_provided_inputs(self, batch):
+
+        return self.__collect_provided_arrays(
+            {k: v for k, v in self.inputs.items()}, batch
```
Comment (pattonw) [inline gunpowder/jax/nodes/train.py:262]: Whats this dict comprehension for?

File: gunpowder/jax/nodes/train.py [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+        )
+
+    def __collect_provided_arrays(self, reference, batch, expect_missing_arrays=False):
```
Comment (pattonw) [inline gunpowder/jax/nodes/train.py:265]: What's the use case for `expect_missing_arrays`? I don't see a way of using it.

File: gunpowder/jax/nodes/train.py [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+        if self.n_devices > 1:
+            for k, v in inputs.items():
+                inputs[k] = jax.device_put_sharded(v, jax.local_devices())
+        else:
+            for k, v in inputs.items():
+                inputs[k] = jax.device_put(v)
+
+        # initialize model if necessary
+        if self.model_params is None:
+            rng = jax.random.PRNGKey(42)
```
Comment (pattonw) [inline gunpowder/jax/nodes/train.py:180]: Should the random key be hard coded like this?

File: gunpowder/jax/nodes/train.py [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+                if self.n_devices > 1:
+                    self.model_params = self.replicate_params(self.model_params)
+        else:
+
+            logger.info("Starting training from scratch")
+            self.model_params = None
+
+    def split_inputs(self, inputs):
+        for k, arr in inputs.items():
+            assert arr.shape[0] % self.n_devices == 0
```
Comment (pattonw) [inline gunpowder/jax/nodes/train.py:153]: This should maybe be documented. I'm guessing shape[0] will usually be the batch dimension. Might be obvious enough that batch size should be a multiple of the number of gpus you have available.
Not super familiar with how this works, can you parallelize computation over any axis? If so might be a weird constraint that some spatial dimension of your input needs to be a multiple of the number of gpus available.

File: gunpowder/jax/nodes/train.py [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+    Args:
+
+        model (subclass of ``gunpowder.jax.GenericJaxModel``):
+
+            The model to train. This model encapsulates the forward model,
+            loss, and optimizer.
+
+        inputs (``dict``, ``string`` -> :class:`ArrayKey`):
+
+            Dictionary from the names of input tensors (argument names of the
```
Comment (pattonw) [inline gunpowder/jax/nodes/train.py:32]: Docstring doesn't seem to match the supported values of inputs that `__collect_provided_arrays` implementation seems to imply. would `Any` -> `Union[str, np.ndarray, ArrayKey]` be more accurate?

File: gunpowder/jax/nodes/train.py [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+            ``train_step`` method) in the model to array keys.
+
+        outputs (``dict``, ``string`` or ``int`` -> :class:`ArrayKey`):
+
+            Dictionary from the names of tensors in the network to array
```
Comment (pattonw) [inline gunpowder/jax/nodes/train.py:37]: Docstring seems mismatched again. Implementation looks like keys here are also `Any`.

File: gunpowder/jax/generic_jax_model.py [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+    def __init__(self):
+        pass
+
+    def initialize(self, rng_key, inputs, is_training):
+        raise RuntimeError("Unimplemented")
+
+    def forward(self, params, inputs):
+        raise RuntimeError("Unimplemented")
+
+    def train_step(self, params, inputs, pmapped=False):
```
Comment (pattonw) [inline gunpowder/jax/generic_jax_model.py:13]: expected outputs should probably be documented. Implementation makes it look like you must return `self.model_params, outputs, loss`. Where outputs is `Dict[Any, np.ndarray]`. I assume loss is just a float and there is some reasonable default storage for jax model parameters?

File: gunpowder/jax/nodes/train.py [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+                if array_key in batch.arrays:
+                    arrays[array_name] = batch.arrays[array_key].data
+                elif not expect_missing_arrays:
+                    logger.warn(msg)
+                else:
+                    logger.debug(msg)
+            elif isinstance(array_key, np.ndarray):
+                arrays[array_name] = array_key
+            elif isinstance(array_key, str):
+                arrays[array_name] = getattr(batch, array_key)
```
Comment (pattonw) [inline gunpowder/jax/nodes/train.py:281]: Should avoid setting string attributes on batches. We've used "loss" and "iteration" like this but that should change and we should avoid allowing any more support for this.
Mainly due to the fact that the gunpowder pipeline often needs to merge 2 batches or create a new batch with a subset of the data in batch. Both of these operations are more annoying with arbitrary string attributes.

File: gunpowder/jax/nodes/__init__.py [hunk 9]
```diff
@@ -0,0 +1,6 @@
+from __future__ import absolute_import
```
Comment (pattonw) [inline gunpowder/jax/nodes/__init__.py:1]: Do we need this? I don't think anyone is still using python2.5

**Review by pattonw — COMMENTED** [review]
I didn't review the Buffered predict yet. I'll make another review when I get around to it.

Overall looks pretty cool. Didn't know parallelization over arbitrary number of gpus could be done so nicely. I'm not particularly familiar with jax (just did a quick google before writing this review) so let me know if I misunderstood something.

File: gunpowder/jax/nodes/train.py (hunk 1 above)
Comment (trivoldus28) [reply gunpowder/jax/nodes/train.py:262]: I think this is for getting items from a batch to a `Dict`. I just basically copied this from the Torch train node: https://github.com/funkey/gunpowder/blob/master/gunpowder/torch/nodes/train.py#L332

Comment (trivoldus28) [reply gunpowder/jax/nodes/train.py:262]: Looking at it I guess it's redundant :sweat_smile: . Will fix it.

File: gunpowder/jax/nodes/train.py (hunk 2 above)
Comment (trivoldus28) [reply gunpowder/jax/nodes/train.py:265]: Same as above, I got this from https://github.com/funkey/gunpowder/blob/master/gunpowder/torch/nodes/train.py#L341. What is the use case of `expect_missing_arrays`? Maybe I can remove it

File: gunpowder/jax/nodes/train.py (hunk 3 above)
Comment (trivoldus28) [reply gunpowder/jax/nodes/train.py:180]: Yea so random key is meant to make training reproducible, but since the Gunpowder pipeline is not deterministic in the first place I guessed it doesn't hurt to hard code the random key for now (and providing an option would be a bit misleading). Can add a comment to make it clearer?

File: gunpowder/jax/nodes/train.py (hunk 4 above)
Comment (trivoldus28) [reply gunpowder/jax/nodes/train.py:153]: For now JAX can only `pmap` across the first dimension. Will add some comment

File: gunpowder/jax/nodes/train.py (hunk 5 above)
Comment (trivoldus28) [reply gunpowder/jax/nodes/train.py:32]: Yup that should be more accurate

File: gunpowder/jax/generic_jax_model.py (hunk 7 above)
Comment (trivoldus28) [reply gunpowder/jax/generic_jax_model.py:13]: Yes will add more documentations

File: gunpowder/jax/nodes/train.py (hunk 8 above)
Comment (trivoldus28) [reply gunpowder/jax/nodes/train.py:281]: Yea I've never used it either; I can remove line 280-281

**Comment by trivoldus28** [comment]
@pattonw I tried to resolve much of your comments... can you take a second look? Thanks!

File: gunpowder/jax/nodes/train.py (hunk 5 above)
Comment (trivoldus28) [reply gunpowder/jax/nodes/train.py]: Thinking about it a little, I think support should only for mapping ``string`` to `ArrayKey` or np.ndarray. Other usages might be possible but unsupported for now. Same for output.

File: gunpowder/jax/nodes/train.py (hunk 1 above)
Comment (pattonw) [reply gunpowder/jax/nodes/train.py]: :+1: I may have written it in the first place if it came from the torch node. Probably better not to propagate it

File: gunpowder/jax/nodes/train.py (hunk 2 above)
Comment (pattonw) [reply gunpowder/jax/nodes/train.py]: I guess for an optional input like loss weights? Maybe there's a good enough use case for arrays that may be provided

File: gunpowder/jax/nodes/train.py (hunk 3 above)
Comment (pattonw) [reply gunpowder/jax/nodes/train.py:186]: Oh ok. It looks like you're targetting the v1.3-dev branch. I built in random seeds to the gunpowder pipeline a little ways back. It might be nice to tie into that. If you look at some of the other nodes that use randomness such as in random location (https://github.com/funkey/gunpowder/blob/4c115bbd944fdc1929427692a9742709c86c12de/gunpowder/nodes/random_location.py#L160) I've been trying to keep random seeds set in the prepare method. This allows requests to be repeated if run with the same seed.
I guess this case is a bit unique since training should change the result of a pipeline request even if the seed used for the request stays the same. But at least if you use the same pipeline with the same initial seed you should get the same first training step.

File: gunpowder/jax/nodes/train.py (hunk 5 above)
Comment (pattonw) [reply gunpowder/jax/nodes/train.py]: Sounds good to me

**Comment by pattonw** [comment]
Documentation is way better, Thanks!

File: gunpowder/nodes/buffered_predict.py [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+from gunpowder.nodes.generic_predict import GenericPredict
+from operator import mul
+
+logger = logging.getLogger(__name__)
+
+
+class BufferedPredict(GenericPredict):
+    '''Buffered implementation of :class:`gunpowder.nodes.Predict`.
+
+    The class runs a background predict process with shared memory buffer
```
Comment (pattonw) [inline gunpowder/nodes/buffered_predict.py:17]: Not quite sure if I get how the BufferedPredict differs from the `spawn_subprocess` flag in GenericPredict.
I may be confused on how the existing predict node works, but I thought the point with that flag was to provide a gpu worker in a seperate process so that we could have multiple workers generating batches and feed them to the predict process. Is the difference just the use of a shared memory buffer instead of the multiprocessing.queue to pass the inputs to the gpu process?

File: gunpowder/nodes/buffered_predict.py [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+        for input_name, input_key in self.inputs.items():
+            if isinstance(input_key, ArrayKey):
+                if input_key in batch.arrays:
+                    inputs[input_name] = batch.arrays[input_key].data
+                else:
+                    logger.warn("batch does not contain %s, input %s will not "
+                                "be set", input_key, input_name)
+            elif isinstance(input_key, np.ndarray):
+                inputs[input_name] = input_key
+            elif isinstance(input_key, str):
```
Comment (pattonw) [inline gunpowder/nodes/buffered_predict.py:244]: another `getattr(batch, string)`

**Review by pattonw — COMMENTED** [review]
The BufferedPredict looks pretty good to me. I would like to see tests for the nodes you added, especially the Jax Train/Predict nodes so we can see an example of how to use them.

File: gunpowder/jax/nodes/train.py (hunk 3 above)
Comment (trivoldus28) [reply gunpowder/jax/nodes/train.py:186]: Nice! Yea I can use that `request.random_seed` for the weight init. One step towards deterministic training :)

Comment (trivoldus28) [reply gunpowder/jax/nodes/train.py:186]: Will the initial seed be the same in the first batch for every run though?

File: gunpowder/nodes/buffered_predict.py (hunk 10 above)
Comment (trivoldus28) [reply gunpowder/nodes/buffered_predict.py:17]: Yea there are a few problems with the current implementation of GenericPredict. First is that it does not allow for multiple workers ([number of workers is always 1](https://github.com/funkey/gunpowder/blob/v1.3-dev/gunpowder/nodes/generic_predict.py#L100)). Increasing the number of workers will lead to some complicated queue reordering problem without locking (worker A/B could get the results B/A from the background predict process). This leads to the current implementation, when using with [daisy predict node](https://github.com/funkey/gunpowder/blob/v1.3-dev/gunpowder/nodes/daisy_request_blocks.py), that spawns a copy of the network model, each with its own weight/memory for each worker which can quickly lead to out of memory problem with multiple workers especially for big models. Also, `multiprocessing.queue` uses `pickle` to serialize data first before queuing which is quite a bit slower than using shared memory.

In short BufferedPredict solves 2 main problems with GenericPredict.
1. Low performance of `multiprocessing.queue`
2. Replicated memory usage of multiple workers

File: gunpowder/nodes/buffered_predict.py (hunk 11 above)
Comment (trivoldus28) [reply gunpowder/nodes/buffered_predict.py]: Removed :)

File: gunpowder/jax/nodes/train.py (hunk 3 above)
Comment (pattonw) [reply gunpowder/jax/nodes/train.py]: The seed can be set when a request is made from the gunpowder pipeline. If the user wants to set the same seed they can, if they want a different seed for multiple random initializations this would make that possible too. I think this looks pretty good now

File: gunpowder/nodes/buffered_predict.py (hunk 10 above)
Comment (pattonw) [reply gunpowder/nodes/buffered_predict.py:17]: I'm surprised that the daisy node + Predict creates separate copies of the model per cpu worker. I don't have an example on me right now but I'm pretty sure I've used these nodes with many workers in the past and it didn't run out of memory. If multiple copies of the model were initialized on the gpu then I wouldn't expect this combo to ever work with more than just a few workers. Also I think setup() should also only be called once since the workers in daisy request blocks reuse the same [upstream provider](https://github.com/funkey/gunpowder/blob/4c115bbd944fdc1929427692a9742709c86c12de/gunpowder/nodes/daisy_request_blocks.py#L133), although here I'm getting a bit further into the details of multiprocessing than I am familiar with and I'm not entirely sure what is referenced and what is copied when creating workers in separate threads/processes.

Improving the performance of the multiprocessing.queue sounds good to me. Do we need a separate node for this? If it reduces memory overhead and speeds up data transfer, it seems like a win win and should just replace the old `GenericPredict` behavior.

**Comment by pattonw** [comment]
Would still like to see tests for the Jax nodes to provide examples on how to use them.

File: gunpowder/nodes/buffered_predict.py (hunk 10 above)
Comment (trivoldus28) [reply gunpowder/nodes/buffered_predict.py:17]: Ah, I think I must have used the Torch predict node without `spawn_subprocess` which may have spawned a copy for each worker. I'll try to rebase the jax predict node to `GenericPredict`. If it works it might be a cleaner add for now and we can work on the shared memory queue later.

Regarding shared memory queue, I'm a bit hesitant replacing multiprocessing.queue completely because it is not as flexible. Currently it only supports transferring shared np.ndarray and pretty much nothing else, so it's probably sufficient for production deployment but not for research/more experimental uses.

Comment (trivoldus28) [reply gunpowder/nodes/buffered_predict.py]: I rebased jax predict to GenericPredict and added some test cases (based on the PyTorch ones). For some reason I couldn't get the multiprocessing tests to not hang with the `Scan` and `PreCache` nodes... maybe there's incompatibilities with jax and multiprocessing and pytest (`PreCache` and `DaisyRequestBlocks` worked fine outside of tests).

**Comment by pattonw** [comment]
Looks good to me now. I think we're good to merge

**Comment by pattonw** [comment]
Actually 1 last thing, can you add the new nodes to the docs? That can be done in the [api.rst](https://github.com/funkey/gunpowder/blob/master/docs/build/api.rst) file

**Comment by trivoldus28** [comment]
Merging to v1.3-dev with @pattonw 's approval.
